File: app/src/main/java/com/paxus/pay/poslinkui/demo/view/AmountTextWatcher.java
package com.paxus.pay.poslinkui.demo.view;

import android.text.Editable;
import android.text.TextWatcher;

import com.pax.us.pay.ui.constant.entry.enumeration.CurrencyType;
import com.paxus.pay.poslinkui.demo.utils.CurrencyUtils;
import com.paxus.pay.poslinkui.demo.utils.Logger;

/**
 * TextWatcher for input format
 * Example: <br>
 * if currency is {@value CurrencyType#USD}<br>
 *      "1" --> "$1.00" <br>
 * if currency is {@value CurrencyType#POINT}<br>
 *      "1" --> "$1.00" <br>
 */
public class AmountTextWatcher implements TextWatcher {

    protected boolean mEditing;
    protected String mPreStr;
    private int maxLength;
    private String currency;
    public AmountTextWatcher(int maxLength, String currency){
        this.maxLength = maxLength;
        this.currency = currency;
    }
    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        if (!mEditing) {
            mPreStr = s.toString();
        }
        Logger.d("AmountTextWatcher beforeTextChanged:"+mPreStr);
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        Logger.d("AmountTextWatcher onTextChanged:"+mPreStr);
    }

    @Override
    public void afterTextChanged(Editable s) {
        if (!mEditing) {
            mEditing = true;
            String value = s.toString().replaceAll("[^0-9]", "");
            if (value.length() == 0) {
                value = "0";
            }
            if(value.length() > maxLength){
                s.replace(0, s.length(), mPreStr);
            } else {
                String formatted = CurrencyUtils.convert(Long.parseLong(value), currency);
                s.replace(0, s.length(), formatted);
                Logger.d("AmountTextWatcher afterTextChanged:"+formatted);
            }
            mEditing = false;
        }
    }
}

File: app/src/main/java/com/paxus/pay/poslinkui/demo/view/AmountTextWatcher.kt
package com.paxus.pay.poslinkui.demo.view

import android.text.Editable
import android.text.TextWatcher
import com.paxus.pay.poslinkui.demo.utils.CurrencyUtils
import com.paxus.pay.poslinkui.demo.utils.Logger

/**
 * TextWatcher for input format
 * Example: <br></br>
 * if currency is {@value CurrencyType#USD}<br></br>
 * "1" --> "$1.00" <br></br>
 * if currency is {@value CurrencyType#POINT}<br></br>
 * "1" --> "$1.00" <br></br>
 */
open class AmountTextWatcher(private val maxLength: Int, private val currency: String?) :
    TextWatcher {
    protected var mEditing: Boolean = false
    protected var mPreStr: String? = null
    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
        if (!mEditing) {
            mPreStr = s.toString()
        }
        Logger.d("AmountTextWatcher beforeTextChanged:" + mPreStr)
    }

    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
        Logger.d("AmountTextWatcher onTextChanged:" + mPreStr)
    }

    override fun afterTextChanged(s: Editable) {
        if (!mEditing) {
            mEditing = true
            var value = s.toString().replace("[^0-9]".toRegex(), "")
            if (value.isEmpty()) {
                value = "0"
            }
            if (value.length > maxLength) {
                s.replace(0, s.length, mPreStr)
            } else {
                val formatted = CurrencyUtils.convert(value.toLong(), currency)
                s.replace(0, s.length, formatted)
                Logger.d("AmountTextWatcher afterTextChanged:" + formatted)
            }
            mEditing = false
        }
    }
}
